docs/streams/developer-guide/processor-api.html

```diff
@@ -396,12 +400,68 @@ <h2>
+                    enables use cases such as temporal stream-table joins.</p>
+                <p>You can create a persistent, versioned state store by passing a
+                    <a class="reference external" href="/{{version}}/javadoc/org/apache/kafka/streams/state/Stores.html#persistentVersionedKeyValueStore(java.lang.String,java.time.Duration)">VersionedBytesStoreSupplier</a>
+                    to the
+                    <a class="reference external" href="/{{version}}/javadoc/org/apache/kafka/streams/state/Stores.html#versionedKeyValueStoreBuilder(java.lang.String,java.time.Duration)">versionedKeyValueStoreBuilder</a>,
+                    or by implementing your own
+                    <a class="reference external" href="/{{version}}/javadoc/org/apache/kafka/streams/state/VersionedKeyValueStore.html">VersionedKeyValueStore</a>.</p>
+                <p>Each versioned store has an associated, fixed-duration <em>history retention</em>
+                    parameter which specifies long old record versions should be kept for.
+                    In particular, a versioned store guarantees to return accurate results for
+                    timestamped retrieval operations where the timestamp being queried is within
+                    history retention of the current observed stream time.</p>
+                <p>History retention also doubles as its <em>grace period</em>, which determines
+                    how far back in time out-of-order writes to the store will be accepted. A
+                    versioned store will not accept writes (inserts, updates, or deletions) if
+                    the timestamp associated with the write is older than the current observed
+                    stream time by more than the grace period. Stream time in this context is
+                    tracked per-partition, rather than per-key, which means it's important
+                    that grace period (i.e., history retention) be set high enough to
+                    accommodate a record with one key arriving out-of-order relative to a
+                    record for another key.</p>
+                <p>Because the memory footprint of versioned key-value stores is higher than
+                    that of non-versioned key-value stores, you may want to adjust your
+                    <a class="reference internal" href="memory-mgmt.html#streams-developer-guide-memory-management-rocksdb"><span class="std std-ref">RocksDB memory settings</span></a>
+                    accordingly. Benchmarking your application with versioned stores is also
+                    advised as performance is expected to be worse than when using non-versioned
+                    stores.</p>
+                <p>Versioned stores do not support caching or interactive queries at this time.
+                    Also, window stores may not be versioned.</p>
+                <b>Upgrade note:</b> Versioned state stores are opt-in only; no automatic
```

Comment:
Should we just refer to the upgrade guide? Or maybe just omit it here at all?

Reply:
Hm, the reason I had it here is because I saw that we had an analogous note for timestamped stores in these docs above: https://github.com/apache/kafka/blob/5afedd9ac37c4d740f47867cfd31eaed15dc542f/docs/streams/developer-guide/processor-api.html#L387

If you think this type of content belongs in the upgrade guide instead, I will move it over (for versioned stores) and simply link to the upgrade guide from here.

Reply:
I've added a section into the upgrade guide which links to the developer guide section. I think this is a good compromise in order to keep the upgrade guide smaller, and to be consistent with the fact that the detailed upgrade note for timestamped key-value stores is also in the developer guide (rather than the upgrade guide).